File: frontend/src/lib/entities/heartbeat.ts
export interface ActiveHours {
	start: string;
	end: string;
	timezone: string;
}

export interface HeartbeatConfig {
	user_id: string;
	assistant_id: string;
	enabled: boolean;
	checklist: string;
	every_seconds: number;
	active_hours: ActiveHours;
	isolated_session: boolean;
	light_context: boolean;
	ack_max_chars: number;
	prompt: string;
	schedule_id: string | null;
}

export interface HeartbeatState {
	last_run_at: string | null;
	last_result: string | null;
	consecutive_ok_count: number;
	next_due_at: string | null;
	total_ticks: number;
	total_escalations: number;
	last_escalation_at: string | null;
}

export interface HeartbeatTickResult {
	action: "ok" | "escalated" | "skipped";
	reason: string;
	response: string | null;
	tokens_used: number | null;
	duration_ms: number | null;
	timestamp: string;
}

export type HeartbeatInterval =
	| 300
	| 900
	| 1800
	| 3600
	| 7200
	| 14400
	| 28800
	| 43200
	| 86400;

export const HEARTBEAT_INTERVALS: {
	value: HeartbeatInterval;
	label: string;
}[] = [
	{ value: 300, label: "5 minutes" },
	{ value: 900, label: "15 minutes" },
	{ value: 1800, label: "30 minutes" },
	{ value: 3600, label: "1 hour" },
	{ value: 7200, label: "2 hours" },
	{ value: 14400, label: "4 hours" },
	{ value: 28800, label: "8 hours" },
	{ value: 43200, label: "12 hours" },
	{ value: 86400, label: "24 hours" },
];

File: backend/src/schemas/entities/heartbeat.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


class ActiveHours(BaseModel):
    start: str = Field(default="09:00", description="Start time in HH:MM format")
    end: str = Field(default="22:00", description="End time in HH:MM format")
    timezone: str = Field(default="UTC", description="IANA timezone identifier")


DEFAULT_HEARTBEAT_PROMPT = (
    "You are a periodic monitoring agent. Review the following checklist and report status.\n\n"
    "Checklist:\n{checklist}\n\n"
    "If everything looks good, respond with HEARTBEAT_OK followed by a brief status (under {ack_max_chars} chars).\n"
    "If any item needs attention, explain the issue in detail."
)


class HeartbeatConfig(BaseModel):
    user_id: str
    assistant_id: str
    enabled: bool = Field(default=False)
    checklist: str = Field(default="", description="Markdown checklist for the heartbeat agent to review")
    every_seconds: int = Field(default=3600, ge=300, le=86400, description="Interval between ticks in seconds")
    active_hours: ActiveHours = Field(default_factory=ActiveHours)
    isolated_session: bool = Field(default=True, description="Run each tick in an isolated chat session")
    light_context: bool = Field(default=True, description="Use minimal context for tick invocations")
    ack_max_chars: int = Field(default=300, description="Max chars allowed after HEARTBEAT_OK token")
    prompt: str = Field(default=DEFAULT_HEARTBEAT_PROMPT, description="Prompt template with {checklist} placeholder")
    schedule_id: Optional[str] = Field(default=None, description="TaskIQ schedule ID when registered")


class HeartbeatState(BaseModel):
    last_run_at: Optional[datetime] = None
    last_result: Optional[str] = None
    consecutive_ok_count: int = Field(default=0)
    next_due_at: Optional[datetime] = None
    total_ticks: int = Field(default=0)
    total_escalations: int = Field(default=0)
    last_escalation_at: Optional[datetime] = Field(default=None, description="For escalation rate limiting")


class HeartbeatTickResult(BaseModel):
    action: Literal["ok", "escalated", "skipped"]
    reason: str
    response: Optional[str] = None
    tokens_used: Optional[int] = None
    duration_ms: Optional[int] = None
    timestamp: datetime


class HeartbeatHistory(BaseModel):
    results: list[HeartbeatTickResult] = Field(default_factory=list)
    max_entries: int = Field(default=100)
